mmdet3d/core/bbox/coders/point_xyzwhlr_bbox_coder.py

```diff
@@ -0,0 +1,106 @@
+import numpy as np
+import torch
+
+from mmdet.core.bbox import BaseBBoxCoder
+from mmdet.core.bbox.builder import BBOX_CODERS
+
+
+@BBOX_CODERS.register_module()
+class PointXYZWHLRBBoxCoder(BaseBBoxCoder):
+    """Point based Bbox Coder for 3D boxes.
+
+    Args:
+        code_size (int): The dimension of boxes to be encoded.
+        use_mean_size (bool): Whether using anchors based on class.
+            Defaults to True
+    """
+
+    def __init__(self, code_size=7, use_mean_size=True, **kwargs):
+        super(PointXYZWHLRBBoxCoder, self).__init__()
+        self.code_size = code_size
+        self.use_mean_size = use_mean_size
+        if self.use_mean_size:
+            self.mean_size = torch.from_numpy(np.array(
+                kwargs['mean_size'])).cuda().float()
+            assert self.mean_size.min() > 0
+
+    def encode(self, gt_bboxes_3d, points, gt_labels_3d=None):
+        """Encode ground truth to prediction targets.
+
+        Args:
+            gt_bboxes_3d(BaseInstance3DBoxes): Ground truth bboxes \
+                with shape (N, 7 + C).
+            points(torch.Tensor): Point cloud with shape(N, 3)
+            gt_labels_3d(torch.Tensor): Ground truth classes
+        Returns:
+            torch.tensor: Encoded boxes with shape(N, 8 + C)
+        """
+
+        gt_bboxes_3d[:, 3:6] = torch.clamp_min(gt_bboxes_3d[:, 3:6], min=1e-5)
+
+        xg, yg, zg, dxg, dyg, dzg, rg, *cgs = torch.split(
+            gt_bboxes_3d, 1, dim=-1)
+        xa, ya, za = torch.split(points, 1, dim=-1)
+
+        if self.use_mean_size:
+            assert gt_labels_3d.max() <= self.mean_size.shape[0] - 1
+            point_anchor_size = self.mean_size[gt_labels_3d]
+            dxa, dya, dza = torch.split(point_anchor_size, 1, dim=-1)
+            diagonal = torch.sqrt(dxa**2 + dya**2)
+            xt = (xg - xa) / diagonal
+            yt = (yg - ya) / diagonal
+            zt = (zg - za) / dza
+            dxt = torch.log(dxg / dxa)
+            dyt = torch.log(dyg / dya)
+            dzt = torch.log(dzg / dza)
+        else:
+            xt = (xg - xa)
+            yt = (yg - ya)
+            zt = (zg - za)
+            dxt = torch.log(dxg)
+            dyt = torch.log(dyg)
+            dzt = torch.log(dzg)
+
+        cts = [g for g in cgs]
```

Comment:
Why not `cts = cgs`? I think this piece of code doesn't clone the tensor so I'm not sure what you want to achieve.

Reply:
We are not sure the cts including what? so i do that. To be honest, I just follow the origin code.

Reply:
change to directly use cgs